cooked-validators/src/Cooked/Attack/Common.hs

```diff
@@ -1,234 +1,200 @@
+--   a /modified transaction/
+--
+-- - /returns/ the value(s) in the 'snd' component of the pairs in the output
+--   list. This is reflected by the 'Monad' instance for 'Attack's.
+newtype Attack a = Attack {getAttack :: MockChainSt -> TxSkel -> [(TxSkel, a)]}
+
+-- | Internal wrapper type fpr compatibility with the LTL modalities. You'll
+-- probably never work with this type if you want to build and use attacks.
+data UntypedAttack where
+  UntypedAttack :: Attack a -> UntypedAttack
+
+instance Functor Attack where
+  fmap f (Attack g) = Attack $ \mcst skel -> second f <$> g mcst skel
+
+instance Applicative Attack where
+  pure x = Attack $ \_ skel -> [(skel, x)]
+  (<*>) = ap
+
+instance Monad Attack where
+  Attack g >>= h = Attack $ \mcst skel ->
+    concatMap (\(skel', x) -> getAttack (h x) mcst skel') $ g mcst skel
+
+instance Alternative Attack where
+  empty = Attack $ \_ _ -> []
+  Attack f <|> Attack g = Attack $ \mcst skel -> f mcst skel ++ g mcst skel
+
+instance MonadPlus Attack
+
+-- * Simple attacks
+
+-- | The never-applicable attack.
+failingAttack :: Attack a
+failingAttack = empty
+
+-- | The attack that always applies and leaves the transaction unchanged.
+doNothingAttack :: Attack ()
+doNothingAttack = return ()
+
+-- | The "attack" that returns the current 'MockChainSt'
+mcstAttack :: Attack MockChainSt
+mcstAttack = Attack $ \mcst skel -> [(skel, mcst)]
+
+-- | The "attack" that obtains some value from the 'TxSkel'.
+viewAttack :: Is k A_Getter => Optic' k is TxSkel a -> Attack a
+viewAttack optic = Attack $ \_mcst skel -> [(skel, view optic skel)]
+
+-- | The attack that sets a certain value in the 'TxSkel'.
+setAttack :: Is k A_Setter => Optic' k is TxSkel a -> a -> Attack ()
+setAttack optic newValue = Attack $ \_mcst skel -> [(set optic newValue skel, ())]
+
+-- | The atack that modifies a certain value in the 'TxSkel'.
+overAttack :: Is k A_Setter => Optic' k is TxSkel a -> (a -> a) -> Attack ()
+overAttack optic change = Attack $ \_mcst skel -> [(over optic change skel, ())]
 
--- * Constructing 'Attack's that return at most one modified transaction
+-- * Composing attacks
 
--- | The simplest way to make an attack from an optic: Try to apply a given
--- function of type @a -> Maybe a@ to all foci of an optic, modifying all foci
--- that return @Just@. This attack returns a singleton list of the modified
--- 'TxSkel' if at least one modification was successful, i.e. if at least one of
--- the foci evaluates to @Just@; otherwise it returns @[]@.
+-- | Turn a potentially failing attack into an always successful attack, that
+-- just leaves the 'TxSkel' unmodified if the original attack would have failed.
+--
+-- In cases where the original attack would have failed, this returns
+-- @Nothing@. If the original attack would have been applicable, this is
+-- signalled by wrapping the original attack's return value in a @Just@.
+skipFailure :: Attack a -> Attack (Maybe a)
```

Comment:
Maybe name this `tryAttack`, in analogous fashion to the `try` function for exceptions.

Reply:
I agree, this is what we had named it in our audit as well